### PR TITLE
feat: Add gemma4 drafter model support

### LINKED ISSUE #2184: MTP support for gemma4

### PR DESCRIPTION
**Pull request by athitten**
# What does this PR do ?

This PR adds joint fine-tuning support for Gemma 4 base (also called target model) and [drafter/assistant models](https://huggingface.co/docs/transformers/en/model_doc/gemma4_assistant) that enable multi-token prediction (MTP). The drafter is co-trained with the Gemma 4 base end-to-end via a composite model (`Gemma4WithDrafter`) that wires up shared K/V states, sqrt(H_b)-scaled embeddings, and a K-step recurrent forward matching the [Gemma 4 drafter tech report](https://x.com/googlegemma/status/2051694045869879749). The PR provides two reference configs for joint fine-tuning of [gemma-4-E4B-it](https://huggingface.co/google/gemma-4-E4B-it) and [gemma-4-E4B-it-assistant](https://huggingface.co/google/gemma-4-E4B-it-assistant), one with MedPix VQA dataset and the other with a text-only Tulu-3 + Magicoder mix. Also provides an inference benchmark script that validates speculative-decode throughput on the saved checkpoint.

The feature has been verified only against the **Gemma 4 4B (E4B)** base + drafter pair. The composite is architecturally model-agnostic within the Gemma 4 family, but the example YAMLs, the parity tests, and all training/inference verification in this PR target the 4B pair.

# Implementation

- **`Gemma4WithDrafter` composite** (`nemo_automodel/components/models/gemma4_drafter/`)
  - Wraps a `Gemma4ForConditionalGeneration` base + `Gemma4AssistantForCausalLM` drafter as a single `nn.Module` for FSDP2 training.
  - K-step recurrent forward: round 0 feeds `cat(base.embed(input_ids), base.h_final)`; round k≥1 feeds `cat(base.embed(input_ids_shifted_by_k), prev_drafter.last_hidden_state)`. `shared_kv_states` is captured **once** from a single base forward and reused at every round (per the Gemma 4 drafter tech report).
  - Saves to `<ckpt>/base/` and `<ckpt>/drafter/` HF-loadable subdirs for vLLM / HF inference handoff.
  - Conditionally unfreezes `post_projection` only when `drafter_num_steps > 1`
  - `masked_embedding.centroids` always frozen (the `torch.topk` inside it blocks gradient flow back to the centroids).

- **Recipe** (`recipes/vlm/finetune.py`, `recipes/base_recipe.py`)
  - New `_maybe_add_drafter_loss(out, base_loss, labels, …)` sums per-step CE over the composite's `drafter_logits` list with `_shift_labels_left(labels, k)` per round.
  - `base_recipe.load_checkpoint` dispatches to `model.load_pretrained` when the composite exposes it, so the `base/` + `drafter/` subdir layout reloads correctly on resume.

- **Text-only dataset adapter** (`components/datasets/vlm/datasets.py`)
  - `make_tulu3_magicoder_text_mix_dataset` interleaves `allenai/tulu-3-sft-mixture` (80 %) and `ise-uiuc/Magicoder-OSS-Instruct-75K` (20 %) into `{"conversation": [...]}` dicts with **no** `image` field, so `default_collate_fn` emits batches without `pixel_values`. The composite + base accept text-only inputs unchanged.

- **Example configs and benchmark** (`examples/vlm_finetune/gemma4_joint_drafter/`)
  - `gemma4_4b_joint_drafter_medpix.yaml` —  Joint fine-tuning recipe of `google/gemma-4-E4B-it` + `google/gemma-4-E4B-it-assistant` with MedPix VQA
  - `gemma4_4b_joint_drafter_tulu_magicoder_mix.yaml` — Joint fine-tuning on tulu and magiccoder mix, same base + drafter pair
  - `benchmark_mtp_inference.py` — measures speculative-decode acceptance + throughput end-to-end against the trained 4B pair
  
 # Testing and Verification
 
 > All numbers below are for joint fine-tuning of `gemma-4-E4B-it` with `/gemma-4-E4B-it-assistant`
 1. **Fine-tuning the joint model on MedPix-VQA: no NaNs, no loss spikes, no grad-norm spikes.** Using the recipe `examples/vlm_finetune/gemma4_joint_drafter/gemma4_4b_joint_drafter_medpix.yaml`
 Loss curve:
Screenshot 2026-05-24 at 4 25 48 PM

2. **Large-scale fine-tuning on a Tulu-3 (80 %) + Magicoder (20 %) mix for 500 steps.** Stable fine-tuning run observed with the recipe `examples/vlm_finetune/gemma4_joint_drafter/gemma4_4b_joint_drafter_tulu_magicoder_mix.yaml`
Loss curve:
Screenshot 2026-05-24 at 4 28 18 PM

3. **Inference run on the tulu + magicoder fine-tuned checkpoint after 500 steps.** Ran `benchmark_mtp_inference.py`(uses transformers generate) against the saved `base/` + `drafter/` pair. Results below:

image

4. **Verified save + standalone HF reload works correctly. Checkpoints are stored in separate `<ckpt>/base/` and `<ckpt>/drafter/` subdirs as a prerequisite for generation.** After a 5-step training run, each sub-checkpoint loads via plain HF `from_pretrained` without any NeMo-specific code:
   - `Gemma4ForConditionalGeneration.from_pretrained("<ckpt>/base/model/consolidated")` → **7.94 B params** ✓
   - `Gemma4AssistantForCausalLM.from_pretrained("<ckpt>/drafter/model/consolidated")` → **78.5 M params** ✓
   - State-dict key sets match the upstream released references ([`google/gemma-4-E4B-it`](https://huggingface.co/google/gemma-4-E4B-it) and [`google/gemma-4-E4B-it-assistant`](https://huggingface.co/google/gemma-4-E4B-it-assistant)) exactly (0 missing, 0 extra).
   - `masked_embedding.token_ordering` (`int64[262144]`) buffer survives the DCP → safetensors path. 
 
 5. **Verified resume-from-checkpoint loss parity.** Run A: 10 steps fresh with `ckpt_every_steps=5`. Run B: same config but `--checkpoint.restore_from <RunA>/epoch_0_step_4`. Per-step loss is **bit-identical** between A and B at every overlapping step (5–9): `3.1257, 3.2992, 3.4045, 2.5375, 3.2824`. Confirms model weights + optimizer state + LR scheduler state + dataloader state + RNG all restore correctly. 
 
 6. **Unit and functional tests.**
   - `tests/unit_tests/models/gemma4_drafter/test_composite.py` — composite pre-projection input layout, shared-KV plumbing, K-step recurrence, save/load round-trip.
   - `tests/unit_tests/models/gemma4_drafter/test_composite_fsdp2.py` — FSDP2 wrap + expert/grad sync.
   - `tests/unit_tests/models/gemma4_drafter/test_drafter_wrapper.py` — drafter sub-module integration.
   - `tests/unit_tests/recipes/test_vlm_drafter_helpers.py` — `_shift_labels_left` + `_maybe_add_drafter_loss`.
   - `tests/functional_tests/L2_HF_Transformer_VLM_Gemma4_Joint_Drafter.sh` — L2 single-node smoke.

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

- Related to # (issue)

### CODE REVIEW

File: docs/model-coverage/vlm/google/gemma4.md
Comment (jgerh) [inline docs/model-coverage/vlm/google/gemma4.md]: We try to avoid latinisms in text, such as "via"

```suggestion
[Gemma 4](https://ai.google.dev/gemma) is Google's next-generation multimodal Gemma family, supporting image-text inputs with a Mixture-of-Experts (MoE) language backbone at larger scales. NeMo AutoModel replaces the HF-native dense matmul over experts with the NeMo `GroupedExperts` backend, enabling Expert Parallelism (EP) through the standard MoE parallelizer.
```

File: docs/model-coverage/vlm/google/gemma4.md [hunk 1]
```diff
@@ -13,14 +13,15 @@
...
-- **Gemma 4 E4B IT** (VL, dense, kv-shared layers)
+- **Gemma 4 E2B IT** (VL, dense, kv-shared layers)
+- **Gemma 4 E4B IT** (VL, dense, kv-shared layers), **Gemma 4 E4B IT Assistant** (Assistant/drafter model for MTP)
 - **Gemma 4 31B IT** (VL, dense)
 - **Gemma 4 26B-A4B IT** (VL, MoE)
 
 ## Architecture
 
 - `Gemma4ForConditionalGeneration`
+- `Gemma4AssistantForCausalLM` (MTP drafter / assistant for speculative decoding; co-trainable with the target Gemma 4 base via `Gemma4WithDrafter`)
```
Comment (jgerh) [inline docs/model-coverage/vlm/google/gemma4.md]:
```suggestion
- `Gemma4AssistantForCausalLM` (MTP drafter / assistant head for speculative decoding; co-trainable with the target Gemma 4 base using `Gemma4WithDrafter`)
```

File: docs/model-coverage/vlm/google/gemma4.md
Comment (jgerh) [inline docs/model-coverage/vlm/google/gemma4.md]:
```suggestion
**2. Navigate to the AutoModel directory** (where the recipes are):
```

**Review by jgerh — COMMENTED** [review]
Completed tech pubs review of docs/model-coverage/vlm/google/gemma4.md and left a few copyedits.

File: examples/vlm_finetune/gemma4_joint_drafter/gemma4_4b_joint_drafter_medpix.yaml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  backend: nccl
+  timeout_minutes: 10
+
+rng:
+  _target_: nemo_automodel.components.training.rng.StatefulRNG
+  seed: 42
+  ranked: true
+
+model:
+  _target_: nemo_automodel.components.models.gemma4_drafter.composite.Gemma4WithDrafter.from_pretrained
```
Comment (akoumpa) [inline examples/vlm_finetune/gemma4_joint_drafter/gemma4_4b_joint_drafter_medpix.yaml:50]: Hi @athitten , is this a hard requirement or we could use an autoclass?

Comment (athitten) [reply examples/vlm_finetune/gemma4_joint_drafter/gemma4_4b_joint_drafter_medpix.yaml:50]: Hi @akoumpa , since this is joint/composite model there is no single HF identifier that the autoclass can dispatch on. Both gemma-4-E4B-it and gemma-4-E4B-it-assistant are separate HF identifiers and repos, hence we have `Gemma4WithDrafter.from_pretrained` calls `NeMoAutoModelForImageTextToText.from_pretrained(base_path, …)` and `NeMoAutoModelForCausalLM.from_pretrained(drafter_path, …)`.

If you want to use the autoclass in the yaml, one option is add a `drafter_path` kwarg recognition in the `NeMoAutoModelForImageTextToText`  and use `NeMoAutoModelForImageTextToText` in the yaml and this will detect drafter path and re-dispatche to Gemma4WithDrafter.from_pretrained. But this will lead to some model specific code in the generic autoclass

Comment (athitten) [reply examples/vlm_finetune/gemma4_joint_drafter/gemma4_4b_joint_drafter_medpix.yaml:50]: Let me know what you think

**Review by akoumpa — DISMISSED** [review]
LGTM

**Review by akoumpa — APPROVED** [review]
